Accessibility: Removes hard-coded columns in home facet lists

ckaz opened the pull request:
Replaces #4231.

* introduces multiple column options for two-columned lists such as 'callnumber-first' with pure CSS breaks (W3C accessibility requirement)
* adds various options for customization such as column-fill, different settings for XS/SM/MD devices and more
* adds customizable visual division between home page facet blocks

ckaz commented:
So, finally, here's the new proposal, branched off `dev`:

1. Until a better solution is found, _we maintain the option that the number of entries is set in `module/VuFind/src/VuFind/ContentBlock/FacetList.php`_ (currently 10 entries,  `$columnSize = 10` , after which a link is inserted, which is misleadingly called "More options ..." but leads to "Advanced Search"). 
3. Since all I could glean from the code is that _only  `callnumber-first` are to be displayed as two columns_, I've simplified my suggestion to _one column for all home facets_ except `callnumber-first` , which gets two columns. The column number is set via SCSS variables (along with a bunch of other settings which may be useful). Using variables allows every library to define their own custom settings. 
The two-column option can be overridden (again using SCSS variables) for devices from 0-420px, 420-767px, 767px-991px and above. Of course if can be defaulted to 3 or more columns, too.
In addition, _other exceptions for displaying home facets in multiple columns could be added_ by changing the  setting in `ContenBlock/FacetList.phtml` (currently: `$cssColumnClass = $field == 'callnumber-first' ? 'two-columns' : 'one-column'; `)
4. The multi-column setting uses CSS column breaks to satisfy W3C accessibility requirements.
5. For safety and variability, there are additional settings defining, for example, height (defaulting to auto), which can be combined with `clip` which defines what will happen if entries overflow the defined height.
6. There is a _customizable visual separation_ now, using a vertical line (with color, width + line type) + padding. All of this is defined via SCSS variables, so it can be set to none, or to custom colors, widths etc. This separating line is only active for device width >= 768px, as the blocks containing the lists are stacked on top of each other for smaller devices

Pls. let me know what you think and whether it is feasible to continue considering this.
Cheers, Claas

demiankatz commented on data/cache/.gitignore:
```diff
@@ -1,2 +0,0 @@
-*
```
This file seems to have been deleted by accident. Do you mind putting it back?

ckaz replied:
Yes, will do.

demiankatz commented on themes/bootstrap5/scss/variables.scss:
```diff
@@ -16,3 +16,21 @@ $nav-link-hover-bg:       $gray-lighter !default;
+$home-facet-list-column-overflow: clip !default; // Defines what's to be done if there are more entries than fit the columns
+$home-facet-list-left-padding: 0 !default; // overwrites default left padding for uls (2rem)
+$home-facet-list-list-style: none !default;
+$home-facet-list-right-padding: 1rem !default;
```
We should probably add an end-of-file line break here for style consistency:

```suggestion
$home-facet-list-right-padding: 1rem !default;

```

ckaz replied:
just re-submitted with the requested changes ....

demiankatz replied:
Thanks! I'll wait and see what others have to say before commenting further myself... but let me know if you need my attention on anything in particular.

demiankatz reviewed (changes requested):
Thanks, @ckaz -- I'll let @crhallberg comment on the actual styles before I dive into that part of the conversation, but I noticed a couple of minor things to clean up that are not related to the core of the logic.

demiankatz commented on data/cache/.gitignore:
```diff
@@ -1,2 +1,2 @@
-*
-!.gitignore
+*
+!.gitignore
```
I'm confused -- this file has been restored, but it's still showing as having diffs. I'm not sure why!

demiankatz replied:
I tried restoring the content myself (hence the last two commits on this branch) but it didn't make a difference.

demiankatz replied:
If this doesn't sort itself out and you'd like more help, please let me know -- I can try rewriting the history here to work around the problem. I bet a rebase and force-push will straighten it out if nothing else does.

ckaz replied:
Hi Demian, just force-pushed after rebase and re-restoring .gitignore -- it has some content now. Pls. let me know if it worked out allright. Cheers!

demiankatz replied:
That's sorted it out -- thanks!

crhallberg commented on themes/bootstrap5/templates/ContentBlock/FacetList.phtml:
```diff
@@ -14,7 +14,11 @@
-          <ul class="home-facet-list">
+          <?php
+            // Special case: two columns for LC call numbers ...
+            $cssColumnClass = $field == 'callnumber-first' ? 'two-columns' : 'one-column'; // Add the CSS class logic here
```
If we're moving the column control to SCSS, we could remove this hard-coding of callnumber-first by adding the field as a class similar to how we handle [page classes](https://github.com/vufind-org/vufind/blob/dev/themes/bootstrap5/templates/layout/layout.phtml#L141).

ckaz replied:
Again, great idea.

crhallberg commented on themes/bootstrap5/scss/components/search.scss:
```diff
@@ -119,12 +119,77 @@ input[type="search"]::-webkit-search-decoration {
+  padding-right: $home-facet-list-right-padding;
+
+  @media (min-width: 768px) {
+    border-right: $home-facet-list-border-right-style; // adds visual division between containers for devices >=768px (containers are stacked on smaller devices)
```
Would [`column-rule`](https://css-tricks.com/almanac/properties/c/column-rule/) be a better fit for this? Would prevent the need for min-width and last-of-type (below). [96% browser availability](https://caniuse.com/mdn-css_properties_column-rule).

ckaz replied:
Sorry for the terrible delay.

Indeed, great idea!

crhallberg commented on themes/bootstrap5/scss/components/search.scss:
```diff
@@ -119,12 +119,77 @@ input[type="search"]::-webkit-search-decoration {
+    }
+
+    @media (min-width: 768px) {
+      flex: 0 1 100%;
```
Should two-column be twice as wide?

ckaz replied:
Not necessarily. It might be useful for some contexts, so if we introduced a variable, this could be adapted very easily.

ckaz reviewed (commented):
Great ideas, many thanks

crhallberg commented:
My only question at this point is if using [column-count](https://developer.mozilla.org/en-US/docs/Web/CSS/column-count) is sufficient, or if we should do something more rigid like [CSS grid](https://developer.mozilla.org/en-US/docs/Web/CSS/grid-template-columns).

ckaz commented:
> My only question at this point is if using [column-count](https://developer.mozilla.org/en-US/docs/Web/CSS/column-count) is sufficient, or if we should do something more rigid like [CSS grid](https://developer.mozilla.org/en-US/docs/Web/CSS/grid-template-columns).

Indeed, this looks tempting but wouldn't it break up the expected flow of the ul lists?
As in this example with `
display: grid;
grid-template-columns: 1fr 1fr;`
added
<img width="727" height="412" alt="grid-template-columns" src="https://github.com/user-attachments/assets/79469615-e5e8-4cfe-ab30-0bb5da1299a5" />

vs. the default version with
`column-count: 2; 
  column-gap: 1rem;
  column-rule: 1px dotted gray;`
  
  
<img width="778" height="451" alt="column-count" src="https://github.com/user-attachments/assets/3b3a2bc6-91de-4a46-8ffa-6a316b1fb196" />

crhallberg commented:
I see! I misunderstood the list layout vs the layout of the facets. My mistake!

demiankatz commented on themes/bootstrap5/templates/ContentBlock/FacetList.phtml:
```diff
@@ -14,6 +14,7 @@
       <div class="home-facet <?=$this->escapeHtmlAttr($field) ?><?=$isHierarchy ? ' facet-tree' : ''?>">
         <h2><?=$labelHeading?></h2>
         <div class="home-facet-container">
+          <?php /* removed the 'callnumber-first' coding here because 'callnumber-first' is set on the 'home-facet' container, so columns can be set using SCSS */ ?>
```
I don't think we need to keep this comment here; if we're removing the hard-coded behavior, we don't need to keep a comment about its removal forever. I think people can rely on git to figure out the history if they're curious about the change.


```suggestion
```

demiankatz replied:
My suggestion about removing this comment was marked as resolved but not applied. I still think we can remove this comment; do you disagree?

demiankatz commented on themes/bootstrap5/templates/ContentBlock/FacetList.phtml:
```diff
@@ -26,10 +27,7 @@
-                // Special case: two columns for LC call numbers...
-                $maxListLength = $field == 'callnumber-first'
-                  ? $columnSize * 2 : $columnSize;
+                $maxListLength = $columnSize;
```
This change causes all the facets to revert to a single-column view. I think we still need a way to allow more values for fields where we want the double-column view. Ideally, it would be nice if the default configuration here resulted in something that looks the same as the default configuration in the current dev branch to demonstrate how to implement flexible columns.

Do you think it would make sense to support field-specific column sizes via the configuration, or am I missing/misunderstanding the point of the work so far?

And just to be clear where I'm coming from, here is how the dev branch looks:

<img width="1680" height="1002" alt="Image" src="https://github.com/user-attachments/assets/6ab6b17b-bc87-4a5a-8fd3-7834bdefb297" />

And here is how this branch looks:

<img width="1680" height="1002" alt="Image" src="https://github.com/user-attachments/assets/3d069012-8170-4ca2-8df7-5a4020dbe90e" />

demiankatz reviewed (changes requested):
Thanks, @ckaz -- see below for some thoughts.

ckaz commented:
> Thanks, @ckaz -- see below for some thoughts.

Hi @demiankatz , thank you. 
Erm, yes and know. I understand you concern, but we should **scrap the original two-column version because it simply is not accessible** according to the WCAG rules: The list must not be split in the hard code html, because it is _one_ list.

I'm on sick leave right now, so I'll look into the issue again in a couple of days. Cheers.

demiankatz commented:
@ckaz, I agree that we should implement the two-column appearance with more accessible markup -- which you have done here. My point is that in order to populate the new markup, we still need to retrieve more values for the facets that are supposed to display in two columns, and I think we need some kind of mechanism to accomplish that.

In any case, I hope you feel better; thanks for taking the time to comment even while not in office (and please don't feel any urgency to respond to this before you return).

ckaz commented:
> @ckaz, I agree that we should implement the two-column appearance with more accessible markup -- which you have done here. My point is that in order to populate the new markup, we still need to retrieve more values for the facets that are supposed to display in two columns, and I think we need some kind of mechanism to accomplish that.

Hi @demiankatz , yes, I am much better ;-)

I'm afraid I don't quite understand what you mean with "we still need to retrieve more values for the facets that are supposed to display in two columns"? 
Do you have any idea, what that mechanism could look like?

demiankatz commented:
> I'm afraid I don't quite understand what you mean with "we still need to retrieve more values for the facets that are supposed to display in two columns"? Do you have any idea, what that mechanism could look like?

There is code in the template that sets a `$maxListLength` variable to determine how many values are displayed. In the old code, this was doubled through a hard-coded check for the call number facet in order to populate two columns. In your revisions, you remove that special case. As a result, we only display one column of call numbers because we stop outputting values before we have used enough to fill the second column. I think we still need a mechanism to configure which fields display twice as many values, or else there is no point in having the multi-column CSS since it will never be used.

ckaz commented:
> > I'm afraid I don't quite understand what you mean with "we still need to retrieve more values for the facets that are supposed to display in two columns"? Do you have any idea, what that mechanism could look like?
> 
> There is code in the template that sets a `$maxListLength` variable to determine how many values are displayed. In the old code, this was doubled through a hard-coded check for the call number facet in order to populate two columns. In your revisions, you remove that special case. As a result, we only display one column of call numbers because we stop outputting values before we have used enough to fill the second column. I think we still need a mechanism to configure which fields display twice as many values, or else there is no point in having the multi-column CSS since it will never be used.

Ok, I hope I get this right: In this PR we have the option to display content of one ul in two columns (conformant to WCAG). So it is possible to display call numbers as a one column or two-column ul (depending on the setting chooses in the SCSS variables `column-count: $home-facet-list-column-count-default`). 
The call number ul gets assigned a class `callnumber-first`, so the value of the SCSS variable outputs it the way we have set.
What am I missing? 
Is there a case where there are so few call numbers that we would not need two columns? On libraries such as that, they can pick the one-column option. Same in cases, where libraries choose to have the call numbers in one column anyway.
I'm probably still oblivious to a problem you see but I cannot for some reason.
Cheers!

demiankatz commented:
> I'm probably still oblivious to a problem you see but I cannot for some reason. Cheers!

@ckaz, in the template code, there is a variable called `$maxListLength` which contrains how many values are displayed in each facet list. In order to display two columns full of values for a particular facet, this variable needs to be doubled; otherwise, even if two columns worth of values exist, they will be truncated before they are displayed. In the old code, we had hard-coded logic to double this value for the call number facet, but that logic has been removed without any replacement -- thus, even though the CSS offers the capability of displaying two columns, the data being used to generate the HTML is always cut short, and only one column will ever be displayed. We either need to restore the old hard-coded logic (maybe an acceptable short-term solution to maintain existing behavior) or else we need to make this configurable (which we could do separately as a follow-up action if it's best to do the short-term fix here).

crhallberg commented on themes/bootstrap5/templates/ContentBlock/FacetList.phtml:
This is the accessibility problem you were pointing out, right?

demiankatz replied:
Good catch, @crhallberg, I think that code was removed in an earlier version of this PR; if it has returned, that's a problem.

ckaz replied:
@crhallberg exactly!

ckaz commented:
> > I'm probably still oblivious to a problem you see but I cannot for some reason. Cheers!
> 
> @ckaz, in the template code, there is a variable called `$maxListLength` which contrains how many values are displayed in each facet list. In order to display two columns full of values for a particular facet, this variable needs to be doubled; otherwise, even if two columns worth of values exist, they will be truncated before they are displayed. In the old code, we had hard-coded logic to double this value for the call number facet, but that logic has been removed without any replacement -- thus, even though the CSS offers the capability of displaying two columns, the data being used to generate the HTML is always cut short, and only one column will ever be displayed. We either need to restore the old hard-coded logic (maybe an acceptable short-term solution to maintain existing behavior) or else we need to make this configurable (which we could do separately as a follow-up action if it's best to do the short-term fix here).

Ok, thanks for clarifying. I think I get it now.

ckaz commented:
@crhallberg and @demiankatz So couldn't we simply, at least as an interim solution, re-instate the doubling of the column content:

`// Special case: two columns for LC call numbers...`
`   $maxListLength = $field == 'callnumber-first'`
`   ? $columnSize * 2 : $columnSize;`

and then remove/rewrite the elseif in lines 53 to 58 so the first `<ul>` isn't closed and no second `<ul>` is opened

` <?php if ($i >= $maxListLength): // list too long? show more link! ?>`
`     <li><a href="<?=$moreUrl?>"><strong><?=$this->transEsc('more_options_ellipsis')?></strong></a></li>`
`      <?php break; ?>`
`    <?php elseif ($i % $columnSize === 0): // end of column? insert break! ?>`
`       </ul><ul class="home-facet-list">`
`  <?php endif; ?>`
                
This would give us the desired number of items in one `<ul>`, which we can then style with the SCSS variables.

demiankatz commented:
Yes, @ckaz, that's more or less what I would suggest, and we can add configurability as a followup later.

ckaz commented:
Sorry for the confusion, @demiankatz . Here's my new suggestion: We re-instate the extra column content for callnumber-first home facet and remove the extra `<ul>` and break the remaining single `<ul>` into columns using CSS.

demiankatz commented on themes/bootstrap5/scss/components/search.scss:
```diff
@@ -103,12 +103,83 @@ input[type="search"]::-webkit-search-decoration {
+      column-count: $home-facet-list-column-count-md; // Defaults to '2'; can be defined in 'variables.scss'
+      column-rule: $home-facet-list-column-rule-md; // Defaults to '1 px solid grey'; can be defined in 'variables.scss'
+    }
+    }
```
There seems to be an indentation problem here; why are these braces on the same level? Is something mismatched somewhere?

demiankatz commented on themes/bootstrap5/scss/components/search.scss:
```diff
@@ -103,12 +103,83 @@ input[type="search"]::-webkit-search-decoration {
+
+.search-home-facets .home-facet:last-of-type {
+  border-right: $home-facet-list-last-of-type-border-right-style; // removes visual division between containers for last container
+  }
```
I don't think this brace should be indented:

```suggestion
}
```

demiankatz reviewed (changes requested):
@ckaz, I think we're on the same page about goals, but things aren't working for me yet. Here's how it looks on the dev branch:

<img width="1920" height="1032" alt="Image" src="https://github.com/user-attachments/assets/e1f33155-c4b4-433d-b5fd-69cdf48384bc" />

Here's how it looks here:

<img width="1920" height="1032" alt="Image" src="https://github.com/user-attachments/assets/c41843df-8a73-4d30-8728-3f89b23e4346" />

...so something is preventing the second column formatting from being applied.

This observation may or may not be related to the problem, but I think it's worth considering in either case:

I notice that your SCSS has hard-coded rules in it related to callnumber-first, but you have variables for "non-default-column-count." I wonder if we can clean this up so the hard-coded assumptions are in fewer places. Since the template is currently the place where we make decisions based on the field name, can we apply a class (like "double-column") in the template and then use that in all the SCSS in place of hard-coded names? That will make it easier to make this configurable in the future, and should also make the SCSS a little shorter/more readable.

I also spotted a few minor formatting details which are listed below.

ckaz commented:
Hi @demiankatz ! You wrote:

> ...so something is preventing the second column formatting from being applied.

Yes, the setting `$home-facet-list-column-fill:` which I will make to default to `balance`, so the `<ul>` should break.

> I notice that your SCSS has hard-coded rules in it related to callnumber-first, but you have variables for "non-default-column-count." I wonder if we can clean this up so the hard-coded assumptions are in fewer places. Since the template is currently the place where we make decisions based on the field name, can we apply a class (like "double-column") in the template and then use that in all the SCSS in place of hard-coded names? That will make it easier to make this configurable in the future, and should also make the SCSS a little shorter/more readable.

You are talking about lines 146 and 152 in search.scss, I believe. The reason for this is that in FacetList.phtml in line 14 we make sure that the field name is added as a class to the `home-facet` class already in place:  

`<div class="home-facet <?=$this->escapeHtmlAttr($field) ?><?=$isHierarchy ? ' facet-tree' : ''?>">`

So, we get divs with `home-facet callnumber-first` or `home-facet language` and the like. I currently fail to see how a revision of that into `home-facet double-column` would not be over-complicating things. Wouldn't that mean we'd have to decide at some other place, which facets get the longer `<ul>` split column treatment? 

As it is, we are already singling out `callnumber-first` as the _only_ special case in line 30 of FacetList.phtml
`
// Special case: two columns for LC call numbers...
$maxListLength = $field == 'callnumber-first'
? $columnSize * 2 : $columnSize;
`                  

Therefore, I see no reason why we should not refer to `.home-facet.callnumber-first` in the SCSS. (Even though I see your concern generally that something like `double-column` would be better for definitions.)

> I also spotted a few minor formatting details which are listed below.

I'll see to that!

Cheers!

P.S. For some reason I fail to find out how to insert quotes from the files directly here in the code, like you do.

demiankatz commented:
> As it is, we are already singling out `callnumber-first` as the _only_ special case in line 30 of FacetList.phtml `// Special case: two columns for LC call numbers... $maxListLength = $field == 'callnumber-first' ? $columnSize * 2 : $columnSize;`
> 
> Therefore, I see no reason why we should not refer to `.home-facet.callnumber-first` in the SCSS. (Even though I see your concern generally that something like `double-column` would be better for definitions.)

My goal, as a follow-up to this PR, is to add a config setting to facets.ini that replaces the hard-coded callnumber-first reference in FacetList.phtml, so that ANY field can be configured to show up as double-rows. A config setting can influence a template, but it cannot dynamically generate SCSS. Therefore, I'm looking for a more generic mechanism, so that if somebody wants to customize the column behavior, they only need to change one thing and not two -- that's why I'm suggesting a more semantic CSS class name, and a more general approach to the implementation of styles. I still think it is valuable to include field-name-specific classes on things to support very targeted local customizations, but I think it is better not to use such specific classes in default styles, because it makes customization more difficult.

> P.S. For some reason I fail to find out how to insert quotes from the files directly here in the code, like you do.

If you go to the "Files changed" tab and click on a line number, then shift-click on another line number, it will select a set of lines. You can then click the little + button to add a comment on those lines. When you do this, there is also a "suggest changes" button in the text editor that can be used to directly propose changes to the code. A bit of a hidden feature, but very useful!

demiankatz reviewed (changes requested):
I'm putting this back into "request changes" mode until we finish our conversation about the CSS class strategy and configurability. Once we are on the same page about that, I'll do another hands-on review, but I don't want to look again until I'm certain things won't change another time. :-)

ckaz commented:
> My goal, as a follow-up to this PR, is to add a config setting to facets.ini that replaces the hard-coded callnumber-first reference in FacetList.phtml, so that ANY field can be configured to show up as double-rows. A config setting can influence a template, but it cannot dynamically generate SCSS. Therefore, I'm looking for a more generic mechanism, so that if somebody wants to customize the column behavior, they only need to change one thing and not two -- that's why I'm suggesting a more semantic CSS class name, and a more general approach to the implementation of styles. I still think it is valuable to include field-name-specific classes on things to support very targeted local customizations, but I think it is better not to use such specific classes in default styles, because it makes customization more difficult.

Ok, that was the piece of information I was missing. Yes of course that makes perfect sense. 
Unfortunately, I'm completely lost as to how to fashion code that takes settings from facets.ini via php into the template.
Do you have any idea how we could still solve this?

demiankatz commented:
> Ok, that was the piece of information I was missing. Yes of course that makes perfect sense. Unfortunately, I'm completely lost as to how to fashion code that takes settings from facets.ini via php into the template. Do you have any idea how we could still solve this?

The `getContext()` method of the [VuFind\ContentBlock\FacetList](https://github.com/vufind-org/vufind/blob/dev/module/VuFind/src/VuFind/ContentBlock/FacetList.php) class provides the variables for the template. More logic could be added there to extract a list of double-column fields from the configuration.

demiankatz committed:
Replace deprecated keyCode property. (#4837)

sambhavp96 committed:
Raise Rector code quality level to 22 (#4836)

demiankatz committed:
Fix bug: focus was not restored after closing lightbox. (#4838)

demiankatz committed:
Upgrade to PHPUnit 11 (#4792)

EreMaijala committed:
Fix enabling of payment button when minimum fee is 0. (#4839)

EreMaijala committed:
SierraRest: Fix payability check. (#4840)

Fixes a mistake that prevented the manualFineDescriptions setting from having effect.

maccabeelevine committed:
Blender: Refactor backend and config names (#4820)

sturkel89 committed:
Updated template to remove colon from List: in Create a List dialog box (#4843)

demiankatz committed:
Fix parameter encoding in browse URLs. (#4743)

- Also adds Mink test coverage for Browse functionality.

sturkel89 committed:
Update country-specific links to Stripe's Privacy Policy in language files (#4844)

demiankatz committed:
Improve icon tool right-to-left handling; remove pointless icon. (#4790)

demiankatz committed:
Fix encoding of filter parameters in BrowseController. (#4846)

ronja-koistinen committed:
Remove SteamHandlerTest::testInvalidFilePath (#4842)

This removes VuFindTest\Log\Handler\SteamHandlerTest::testInvalidFilePath
because it was unportable, failed when PHP is run as root (e.g. in a
container), and didn't test any relevant VuFind functionality.

demiankatz committed:
Deprecate support for Memcache extension. (#4814)

demiankatz committed:
Improve OpenAPI search description. (#4848)

demiankatz committed:
Deduplicate facet values when building channels from records (#4847)

sturkel89 committed:
Updated language strings for Spanish courtesy of Jorge Ramírez (#4850)

sturkel89 committed:
Update dev dependencies for 11.0 release - 20251112 (#4851)

sturkel89 committed:
Updated laminas-form, laminas-paginator, swagger-ui, and stripe-php (#4852)

sturkel89 committed:
Updated language files for Czech courtesy of Josef Moravec (#4853)

AnnaSmaros committed:
Update Swedish translations (#4749)

EreMaijala committed:
Fix doc comment for RoleInterface in User entity. (#4854)

demiankatz committed:
Update Vagrant configuration to use Ubuntu 24. (#4849)

EreMaijala committed:
Add resource getter to Comments entity and clean up. (#4856)

Switches the order of getters/setters around and updates descriptions.

EreMaijala committed:
Fix association mappings for extended entities. (#4855)

Fixes an issue where searching by association did not work properly when an entity with associations was extended (e.g. UserList when trying to find one by user).

maccabeelevine committed:
Overdrive: Fix HTML and spacing for status (#4858)

EreMaijala committed:
Add support for index-specific HTTP settings. (#4860)

This makes it easy to e.g. set up SSL connection to Solr using self-signed certificates without affecting other HTTP requests.

EreMaijala committed:
Update Solr schemas to version 1.7. (#4825)

- Enables docValues by default for supporting fields.
- textFacet fields remain uninvertible text fields for the analysis chain to be effective.
- Classes replaced with names where possible.

crhallberg committed:
New Channels: Remove carousel (#4557)

demiankatz committed:
Fix HTML validation issue -- invalid aria-disabled value.

stweil committed:
Fix some German translations (#4865)

Signed-off-by: Stefan Weil <sw@weilnetz.de>

EreMaijala committed:
Admin: Fix recipient info in payment fees list. (#4864)

sturkel89 committed:
Updated more Catalán translations, courtesy of Hugo Agud (#4867)

rajaro committed:
Fix method timed blocks near midnight (#4866)

demiankatz committed:
SimilarItems channel provider: encode IDs in tokens. (#4859)

demiankatz committed:
Fix and simplify noajax handling in embedded tab code. (#4862)

sturkel89 committed:
Updated language files for Italian courtesy of Rodolfo Marraffa (#4869)

sturkel89 committed:
Added translation string for Croatian courtesy of Milo Ivir (#4870)

demiankatz committed:
FOLIO: tolerate missing request preferences. (#4863)

demiankatz committed:
Replace SimilarItemsCarousel tab with Channels tab. (#4209)

damien-git committed:
Improved UI to exclude facets (#4562)

sturkel89 committed:
Updated localization for Stripe URL in ca.ini and it.ini (#4872)

EreMaijala committed:
Use defaultCurrency as default for payment service, improve docs. (#4874)

These changes make it more obvious that the currency codes should match. Otherwise what's displayed would not match the actual amount.

maccabeelevine committed:
Fix default Solr spellchecker thresholdTokenFrequency (#4873)

sturkel89 committed:
Updated Javascript dependencies for 11.0 release (#4871)

dependabot[bot] committed:
Bump js-yaml from 3.14.1 to 3.14.2 (#4875)

Bumps [js-yaml](https://github.com/nodeca/js-yaml) from 3.14.1 to 3.14.2.
- [Changelog](https://github.com/nodeca/js-yaml/blob/master/CHANGELOG.md)
- [Commits](https://github.com/nodeca/js-yaml/compare/3.14.1...3.14.2)

---
updated-dependencies:
- dependency-name: js-yaml
  dependency-version: 3.14.2
  dependency-type: indirect
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

demiankatz committed:
Fix bug: do not double-encode location text. (#4877)

demiankatz committed:
[VUFIND-1664] Improve numbering consistency in Search and Account record lists (#4677)

demiankatz committed:
Solr: add --user-managed switch to eliminate warning. (#4878)

crhallberg committed:
fix: HTML validation fixes for checkboxes (#4879)

demiankatz committed:
Add missing en-gb language subdomain files.

demiankatz committed:
Bump version for 11.0 release.

EreMaijala committed:
Fix online payment email alerts. (#4885)

demiankatz committed:
Fix bug: checkboxes do not disappear in mobile favorites. (#4886)

EreMaijala committed:
Paytrail: fix payment request when there are no row mappings. (#4887)

demiankatz committed:
Standardize select all buttons and account action toolbars.

demiankatz committed:
Fix alignment between select-all checkbox and table content.

demiankatz committed:
Bump version on package-lock.json.

EreMaijala committed:
Adjust table-response margins to avoid content shifting.

EreMaijala committed:
Fix Finnish translation for sort_created_desc. (#4892)

EreMaijala committed:
Fix a HTML validation issue in user ratings list. (#4893)

demiankatz committed:
Fix bug: select all control broken in book bag outside lightbox. (#4889)

EreMaijala committed:
Fix Logger so that it implements the ExtendedLoggerInterface. (#4894)

crhallberg committed:
Checkbox Alignment Improvements (#4883)

Fixes are primarily focused on user content editing and the book bag.

demiankatz committed:
Fix broken tests.

EreMaijala committed:
Trivial: Tweak JS comment style.

LuomaJuha committed:
Throw error on jsdoc (#4897)

EreMaijala committed:
Paytrail: Add product for service fee only if product code specified. (#4898)

dltj committed:
Remove determination of auth Solr core in import-marc-auth script (#4895)

EreMaijala committed:
Fix referrer check in email action. (#4899)

EreMaijala committed:
Avoid potential use of null as array in controller plugins. (#4900)

EreMaijala committed:
Fix error message handling in payment registration. (#4902)

sturkel89 committed:
Updated German translation of 'Holds and Recalls' (#4904)

- Updated translation of long label so that words wrap in My Account menu

maccabeelevine committed:
Fix AbstractExpireCommand handling of float age (#4905)

stweil committed:
Lock lm-commons/lmc-rbac to release 2.1.1 (#4909)

Release 2.2.0 changes the interface of AssertionInterface::assert.

Signed-off-by: Stefan Weil <sw@weilnetz.de>

EreMaijala committed:
Fix exception logging to include the location causing the exception. (#4913)

EreMaijala committed:
Fix database migration sorting by version. (#4911)

EreMaijala committed:
Fix audit event data JSON handling and caller method determination. (#4912)

EreMaijala committed:
Add a space between exception message and server details. (#4915)

demiankatz committed:
Remove obsolete references to Google Analytics universal setting. (#4916)

stweil committed:
Update LDAP authentication to support a full LDAP URI (#4645)

Setting `port` is no longer required in the configuration.
It is ignored if `host` specifies a full LDAP URI.

As of PHP 8.3.0, calling ldap_search with $host and $port is deprecated
(see https://www.php.net/manual/en/function.ldap-connect.php).

Signed-off-by: Stefan Weil <sw@weilnetz.de>
Co-authored-by: Demian Katz <demian.katz@villanova.edu>

demiankatz committed:
Minor autocomplete improvements/modernizations. (#4914)

EreMaijala committed:
Clone any DateTime to return in entities' getters. (#4920)

This avoids indirect/accidental updates to entities if the returned dates are manipulated.

demiankatz committed:
Realign CONTRIBUTING.md with wiki content.

Added links for developer tools and PR reviewing.

EreMaijala committed:
Allow more fine-grained logging level configuration. (#4921)

maccabeelevine committed:
EDS: Fallback to VuFind's cover loader (#4309)

demiankatz committed:
Aleph driver: ensure XML is properly converted to string. (#4925)

maccabeelevine committed:
EDS: Treat 'News', 'Conference' format as also an 'Article' for OpenURL (#4891)

* EDS: Treat 'News' format as also an 'Article' for OpenURL

* Add 'conference' pub type

* Alphabetize

demiankatz committed:
Autocomplete: add setting to apply active filters. (#4246)

dltj committed:
Url-encode username/password parts of DSN (#4910)

Co-authored-by: Ere Maijala <ere.maijala@helsinki.fi>

demiankatz committed:
Fix misplaced parenthesis.

demiankatz committed:
Rename VuFind\Mailer\Factory to MailerFactory for consistency. (#4932)

demiankatz committed:
Standardize PHPdoc to use bool instead of boolean. (#4931)

Includes test to prevent regressions.

Co-authored-by: Thomas Wagener <wagener@hebis.uni-frankfurt.de>

ThoWagen committed:
Add parent_config_name directive to YAML reader (#4930)

stweil committed:
Update GitHub actions (#4936)

demiankatz committed:
Channels tab: enable AJAX compatibility. (#4938)

stweil committed:
Fix German translations for 'website' (#4935)

Signed-off-by: Stefan Weil <sw@weilnetz.de>

crhallberg committed:
Fix: Collection Containment. (#4937)

Co-authored-by: Demian Katz <demian.katz@villanova.edu>

ThoWagen committed:
Fix accessibility of record tabs in v11 (#4903)

demiankatz committed:
Fix incorrect hierarchy field names in comments. (#4940)

sambhavp96 committed:
Add RemoveExpectAnyFromMockRector rule to rector (#4948)

damien-git committed:
Folio: Avoid duplicate bound-with items (#4927)

demiankatz committed:
Fix fund ID type inconsistencies in ILS drivers (mostly comments). (#4949)

EreMaijala committed:
Fix update_resource_metadata util to clear entities after each iteration. (#4950)

This ensures more thoroughly than the previous mechanism that tracked entities do not accumulate and cause excessive memory usage.

It also improves test coverage and removes a useless (and potentially problematic) assertion.

damien-git committed:
[VUFIND-1808] No fancy quote normalization (#4908)

sambhavp96 committed:
Add GetMockBuilderGetMockToCreateMockRector rule to rector (#4951)

demiankatz committed:
Fix bug: inappropriate multiselect options in facet lightbox. (#4955)

stweil committed:
Add more CSP directives to contentsecuritypolicy.ini, and cross-reference in config.ini (#4952)

Signed-off-by: Stefan Weil <sw@weilnetz.de>

demiankatz committed:
Ensure consistent string type on IDs in GetItemStatuses. (#4960)

Co-authored-by: Michael Birkner-Tröger <michael.birkner@akwien.at>

demiankatz committed:
Add Mink test coverage for ILS-driven New Items. (#4947)

demiankatz committed:
Fix embedded record tabs memory bug; add test coverage. (#4961)

maccabeelevine committed:
LibGuides: Exclude hidden AZ databases from connector (#4958)

betullam committed:
Add setting to control holdings tab behavior when no ILS items exist (#4809)

sambhavp96 committed:
Refactor AbstractExpireCommandTest as an abstract TestCase (#4965)

ThoWagen committed:
Allow hashed passwords for PasswordAccess (#4964)

sambhavp96 committed:
Add AssertEqualsToSameRector rule to rector (#4966)

sambhavp96 committed:
Add UseSpecificWithMethod rule to rector (#4967)

EreMaijala committed:
Add support for autowiring of services; apply to OAuth2 code (#4929)

demiankatz committed:
Introduce AutowireableTrait; adjust eligibility criteria. (#4969)

sambhavp96 committed:
Add YieldDataProviderRector rule to rector (#4970)

sambhavp96 committed:
Add AssertInstanceOfComparison rule to rector (#4972)

steinm committed:
Highlight query in autocompleted terms disregarding case (#4963)

sambhavp96 committed:
Add more Rector PHPUnit code quality to rector (#4974)

stweil committed:
Create cache directories for cover images only when needed (#4830)

This fixes a potential race condition which could result in a
runtime warning ("Warning: mkdir(): File exists") on a new
VuFind installation.

The cache directories are only needed when caching is enabled.

It is sufficient to check whether they exist and to create them
when a cover image is written to the cache.

Signed-off-by: Stefan Weil <sw@weilnetz.de>

demiankatz committed:
Add generateRandomHoldings setting to Demo driver. (#4975)

Co-authored-by: Robby <roudonro@msu.edu>

EreMaijala committed:
Fix doc comment for checkMethodregisterPayment. (#4979)

Yihuki committed:
Fix BrowZine coverimage load for Primo backend. (#4720)

Primo backend only returns isbn but not issn, browzine coverimage loader needs issn to get the cover image.

stweil committed:
Fix documentation of return types for Alma ILS driver (#4977)

Signed-off-by: Stefan Weil <sw@weilnetz.de>

demiankatz committed:
OpenLibrarySubjects: minor fixes/style improvements. (#4926)

EreMaijala committed:
Avoid overriding stored Results object by null in RecordLinker. (#4980)

The new resetStoredResults method can be used to reset results if necessary.

maccabeelevine committed:
Blender: Add Blender2 backend to use separate config (#4845)

stweil committed:
Simplify ALMA ILS driver (#4978)

It's not necessary to start the path parameter with the baseUrl when
calling function makeRequest because the baseUrl is added there.

Signed-off-by: Stefan Weil <sw@weilnetz.de>

betullam committed:
Add password hasher to AlmaDatabase (#4982)

maccabeelevine committed:
Add ServerUrl helper (#4983)

dltj committed:
Refactor out holdings-no-items data creation (#4984)

betullam committed:
Fix check request is valid (#4985)

demiankatz committed:
Trivial: fix alphabetical order in config.

EreMaijala committed:
Trivial: Fix a repeated mistake in a comment.

EreMaijala committed:
Fix hold pickup location label handling and title label. (#4988)

dependabot[bot] committed:
Bump qs from 6.12.0 to 6.14.1 (#4989)

Bumps [qs](https://github.com/ljharb/qs) from 6.12.0 to 6.14.1.
- [Changelog](https://github.com/ljharb/qs/blob/main/CHANGELOG.md)
- [Commits](https://github.com/ljharb/qs/compare/v6.12.0...v6.14.1)

---
updated-dependencies:
- dependency-name: qs
  dependency-version: 6.14.1
  dependency-type: indirect
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

demiankatz committed:
Improve exception message for value setting failure (#4990)

dmj committed:
Tolerate multi-valued Solr title fields for broader compatibility (#3279)

The German search index K10plus-Zentral stores title, title_sub, and
title_short as multi-valued. This change to RecordDriver/DefaultRecord
modifies the respective getters for compatibility.

maccabeelevine committed:
Remove result-links empty line (#4986)

demiankatz committed:
Fix minor issues with GetRecordVersions. (#4992)

EreMaijala committed:
Fix AuthHashService::getLatestBySessionId. (#4994)

The method failed when there were multiple rows for the session in the database.

demiankatz committed:
Avoid potential errors from getOneOrNullResult (#4995)

demiankatz committed:
Refactor QueryBuilderTest::testNormalization to use data provider. (#4981)

demiankatz committed:
[VUFIND-1813] Fix facet lightbox for new items. (#4956)

demiankatz committed:
Fix broken SolrJ symlinks in import script. (#4934)

demiankatz committed:
Fix bug: AJAX covers broken in grid view. (#4993)

- Thanks to Ján Krištof for reporting the problem and providing the solution.

aleksip committed:
Add new section service and refactor navigation plugins (#4796)

sambhavp96 committed:
Add PHPUnit Code Quality set to rector (#4998)

stweil committed:
Fix a typo in a lightbox.js comment (#4999)

Signed-off-by: Stefan Weil <sw@weilnetz.de>

cwolfebsco committed:
Folio.ini: Add note about potential account collisions (#4997)

Adding a note to warn about potential account collisions when a non unique username_field such as personal.lastName is selected

Co-authored-by: Demian Katz <demian.katz@villanova.edu>

sambhavp96 committed:
Simplify code using findCss/unFindCss helper methods (#5000)

stweil committed:
Fix AbstractBackgroundLayer::createSeed for empty arguments (#5003)

The function calculated a seed 0 if both arguments were empty strings
(which should not happen normally, but can happen in special cases).

The resulting pattern "0000" then triggered an exception in
GridBackground::renderGrid:
Whoops\Exception\ErrorException, Undefined array key 4.

Because of this exception, no cover image was created and the text
"Cover Image" (or its translation) was shown.

Signed-off-by: Stefan Weil <sw@weilnetz.de>
Co-authored-by: Demian Katz <demian.katz@villanova.edu>

EreMaijala committed:
Add audit event for password reset. (#5007)

maccabeelevine committed:
FOLIO: Ignore allowed service points for delivery requests (#5010)

demiankatz committed:
Improve Mink test types; remove unnecessary assertions. (#5006)

demiankatz committed:
Raise Rector level to 23. (#5014)

demiankatz committed:
Bump version to 11.0.1.

EreMaijala committed:
Drop term "SFX" from Northern Sámi translation for More options. (#5015)

maccabeelevine committed:
Add default for ShardPreferences->showCheckboxes (#5016)

stweil committed:
Fix OpenIDConnect (replace deprecated logout by getLogoutRedirectUrl) (#5018)

- Also includes fix to related whitespace problem in Auth Manager error message.

Signed-off-by: Stefan Weil <sw@weilnetz.de>
Co-authored-by: Demian Katz <demian.katz@villanova.edu>

EreMaijala committed:
Include all source directories in PHPUnit configuration. (#5019)

This allows the coverage report to include all the modules that would be included in tests as well.

maccabeelevine committed:
Explain: remove unused params (#5017)

stweil committed:
Add comments to help configure email on typical servers (#5020)

By default, Debian GNU Linux installs mail software with a self-signed
certificate. This causes a runtime exception when trying to send
feedback emails.

Improve the config.ini documentation to show how certificate
verification can be disabled to enable email sending.

Signed-off-by: Stefan Weil <sw@weilnetz.de>

demiankatz committed:
SearchFacetsTest: add retry to reduce chances of failure. (#5021)

EreMaijala committed:
Trivial: Fix mistake in config.ini comment.

aleksip committed:
Restore @parent_yaml and @parent_config_name false value functionality (#5022)

EreMaijala committed:
Fix record page filter behavior with retention disabled. (#5023)

EreMaijala committed:
Blender: Fix facet DefaultValue with multiple values for a field. (#5024)

EreMaijala committed:
Fix an explodeSetting call in SierraRest. (#5026)

dependabot[bot] committed:
Bump lodash from 4.17.21 to 4.17.23 (#5028)

Bumps [lodash](https://github.com/lodash/lodash) from 4.17.21 to 4.17.23.
- [Release notes](https://github.com/lodash/lodash/releases)
- [Commits](https://github.com/lodash/lodash/compare/4.17.21...4.17.23)

---
updated-dependencies:
- dependency-name: lodash
  dependency-version: 4.17.23
  dependency-type: indirect
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

demiankatz committed:
Better exception if autowiring parameter resolution fails. (#5027)

aleksip committed:
Refactor header bar and footer menu to be configurable section plugins (#5002)

maccabeelevine committed:
EDS: Allow limiter-only queries (#4976)

aleksip committed:
Fix FooterMenu default configuration backward compatibility (#5031)

stweil committed:
Fix IpAddressUtils::normalizeIp and add related tests (#5030)

Signed-off-by: Stefan Weil <sw@weilnetz.de>

demiankatz committed:
Fix New Items HTML (BS5 radio button group markup / button tab order). (#5032)

demiankatz committed:
Update SMS icons to use mobile instead of handset. (#5033)

EreMaijala committed:
Add support for LinkIQ OpenURL resolver. (#5009)

maccabeelevine committed:
BrowZine: Allow disabling individual bestIntegratorLink types (#4774)

dependabot[bot] committed:
Bump phpunit/phpunit from 11.5.43 to 11.5.50 (#5037)

Bumps [phpunit/phpunit](https://github.com/sebastianbergmann/phpunit) from 11.5.43 to 11.5.50.
- [Release notes](https://github.com/sebastianbergmann/phpunit/releases)
- [Changelog](https://github.com/sebastianbergmann/phpunit/blob/11.5.50/ChangeLog-11.5.md)
- [Commits](https://github.com/sebastianbergmann/phpunit/compare/11.5.43...11.5.50)

---
updated-dependencies:
- dependency-name: phpunit/phpunit
  dependency-version: 11.5.50
  dependency-type: direct:development
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

EreMaijala committed:
Update Doctrine and other dependencies. (#5038)

aleksip committed:
Update menu tests and configuration file comments (#5012)

meganschanz committed:
Add in a fallback API call for the Folio driver getModuleMajorVersion function (#5001)

maccabeelevine committed:
Fix DefaultParametersListener and add Mink test (#5013)

demiankatz committed:
Rector fix.

demiankatz committed:
Eliminate bottom border on table headers in mobile view. (#5041)

dependabot[bot] committed:
Bump symfony/process from 7.4.4 to 7.4.5 (#5043)

Bumps [symfony/process](https://github.com/symfony/process) from 7.4.4 to 7.4.5.
- [Release notes](https://github.com/symfony/process/releases)
- [Changelog](https://github.com/symfony/process/blob/8.1/CHANGELOG.md)
- [Commits](https://github.com/symfony/process/compare/v7.4.4...v7.4.5)

---
updated-dependencies:
- dependency-name: symfony/process
  dependency-version: 7.4.5
  dependency-type: indirect
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

demiankatz committed:
Improve open lightbox/click test logic with helper method. (#5044)

aleksip committed:
Fix default section configuration tests (#5045)

demiankatz committed:
Clear local storage between tests to prevent side effects. (#5046)

maccabeelevine committed:
Refactor the DefaultParametersListener test for readability (#5042)

ckaz commented:
Hi Demian,
here's a fresh attempt at making this more configurable as I understood your comment. Please bear with me as I am not at all well-versed as regards PHP. 
My best, Claas

demiankatz committed:
Virtua driver DB improvements: lazy instantiation, better exception (#5029)

demiankatz committed:
Add option of configuring additional scopes for PAIA (#4945)

Co-authored-by: Demian Katz <demian.katz@villanova.edu>
Co-authored-by: Nass, William <william.nass@sbb.spk-berlin.de>
Co-authored-by: William Nass <56260490+Willy8-Dev@users.noreply.github.com>

stweil committed:
Fix Primo backend for dedupmrg records (#5035)

Signed-off-by: Stefan Weil <sw@weilnetz.de>
Co-authored-by: Stefan Weil <stefan.weil@uni-mannheim.de>

demiankatz commented on themes/bootstrap5/templates/ContentBlock/FacetList.phtml:
```diff
@@ -27,9 +30,9 @@
-                $maxListLength = $field == 'callnumber-first'
-                  ? $columnSize * 2 : $columnSize;
+                // Two-column facets (configured in facets.ini: [HomePage_Settings] two_column_facets):
+                //$useTwoColumns = in_array($field, $twoColumnFacets ?? [], true);
```
Is this supposed to be commented out? I suspect not:


```suggestion
                $useTwoColumns = in_array($field, $twoColumnFacets ?? [], true);
```

ckaz replied:
Yes, you are right. Thanks

demiankatz commented on config/vufind/facets.ini:
```diff
@@ -328,7 +328,7 @@ translated_facets[] = language:ISO639-3
 language         = Language
 format           = Format
-;hierarchy_top_title   = Collections
+hierarchy_top_title   = Collections
```
I think this was uncommented by accident:

```suggestion
;hierarchy_top_title   = Collections
```

demiankatz commented on module/VuFind/src/VuFind/ContentBlock/FacetList.php:
```diff
@@ -138,6 +172,7 @@ public function getContext()
             'hierarchicalFacets' => $this->getHierarchicalFacets($facetConfig),
             'hierarchicalFacetSortOptions' =>
                 $this->getHierarchicalFacetSortSettings($facetConfig),
+            'twoColumnFacets' => $this->getTwoColumnFacets($facetConfig),
```
I believe that you can simplify this by adding:

```
use VuFind\Config\Feature\ExplodeSettingTrait;
```

at the top of the class definition, and then changing this line to:


```suggestion
            'twoColumnFacets' => $this->explodeListSetting($facetConfig->HomePage_Settings->two_column_facets ?? ''),
```

This existing trait does much the same thing as your getTwoColumnFacets method, so we might as well use the standard logic instead of creating a new variation.

demiankatz reviewed (changes requested):
Thanks, @ckaz, this does indeed look like what I had in mind; there are just a few details to adjust (see below).

Also, I notice there is a merge conflict here that can likely be resolved by merging the dev branch, rebuilding the CSS, and committing the resulting files to clean up the conflict. If you need any help sorting that out, let me know and I'd be happy to take care of it for you.

I haven't done hands-on testing here yet, but I will once this review has been addressed (or if you need my help with any additional parts of it).

aleksip committed:
Add SiteMap section plugin and controller (#5040)

demiankatz committed:
Add new combined search box setting: collapseInactiveBackendOptions (#4957)

aleksip committed:
Disable inheritance in test data. (#5054)

Co-authored-by: Demian Katz <demian.katz@villanova.edu>

ThoWagen committed:
Add configurable notices (#5048)

ThoWagen committed:
Add configurable notices (#5048)

EreMaijala committed:
Fix LocaleSettingsAwareTrait name. (#5056)

aleksip committed:
Fix SiteMap section plugin templates (#5057)

demiankatz committed:
Improved SiteMap Mink test. (#5055)

maccabeelevine committed:
Add RouteHelper, wrapping Laminas URL helper (#5049)

demiankatz committed:
Remove orphaned dbUpgrade controller plugin configs. (#5060)

EreMaijala committed:
Clean up FlashMessenger calls. (#5058)

maccabeelevine committed:
Allow end-positioning a drop-down menu (#5061)

EreMaijala committed:
Defer creation of URL helper in RouteHelper to avoid early plugin invokation. (#5062)

EreMaijala committed:
Allow the URL helper to be created even if Application does not yet have MvcEvent. (#5063)

While this shouldn't happen, anything (such as an event listener) that tries to create the plugin before MvcEvent is created for Application would cause an error.

dependabot[bot] committed:
Bump qs from 6.14.1 to 6.14.2 (#5070)

Bumps [qs](https://github.com/ljharb/qs) from 6.14.1 to 6.14.2.
- [Changelog](https://github.com/ljharb/qs/blob/main/CHANGELOG.md)
- [Commits](https://github.com/ljharb/qs/compare/v6.14.1...v6.14.2)

---
updated-dependencies:
- dependency-name: qs
  dependency-version: 6.14.2
  dependency-type: indirect
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

aleksip committed:
Add submenuItems support to HeaderBar (#5068)

aleksip committed:
Use YAML in getDefaultMenuConfig methods (#5077)

aleksip committed:
Remove AccountMenu helper dependency (#5076)

ThoWagen committed:
Add content types to notices (#5072)

ThoWagen committed:
Fix HTML validation for notices (#5079)

demiankatz committed:
Upgrade firebase/php-jwt (#5082).

ThoWagen committed:
Hide empty cover links (#5083)

berndoberknapp committed:
Use empty string as default for null password in Folio getUserWithCql (#5088)

aleksip committed:
Finnish, Swedish and Northern Sámi translations for new site map related strings (#5090)

aleksip committed:
Section plugin related improvements (#5089)

stweil committed:
Use standard text for Primo search in all fields (#5071)

Signed-off-by: Stefan Weil <sw@weilnetz.de>

demiankatz committed:
import-marc.sh: always initialize REGENERATE_SOLRJ_DIR (#5078)

EreMaijala committed:
Update current URL when switching switching language or theme. (#5093)

Also applies to print button. The current window URL is modified in JS instead of using the link when the page was created to ensure that it's up to date (e.g. includes latest search result parameters).

demiankatz committed:
Apply Rector.

EreMaijala committed:
Make ListViewsTest more robust. (#5094)

maccabeelevine committed:
RateLimiter: Fix url parameters lost when passing through Turnstile (#5098)

demiankatz committed:
sandal5: fix search home height. (#5099)

ekaterinazinchenko committed:
Allow local sort definitions in the "Sort By" menu (#5091)

These aliases can be configured in the LocalSortDefinitions section of the search configuration file.

LuomaJuha committed:
Null coalesce getCatUsername result in line 295 (#5105)

demiankatz committed:
create_language_translations.js: use absolute path everywhere. (#5101)

demiankatz committed:
Updated Hindi 'zoom' translation, courtesy of Hariharan G. (#5104)

demiankatz committed:
Deprecate ILS-based new item functionality. (#5051)

demiankatz committed:
Fix precedence error in new item config upgrade logic. (#5107)

stweil committed:
Add CSP directive for images from Commons to contentsecuritypolicy.ini, and cross-reference in config.ini (#5113)

Signed-off-by: Stefan Weil <sw@weilnetz.de>

ThoWagen committed:
Improve virtual keyboard selection (#5110)

sturkel89 committed:
Add EDS Advanced Search limiter options and language strings (#5114)

demiankatz committed:
ApiKey entity: Fix typo in PHPdoc. (#5116)

stweil committed:
Update some packages with vulnerabilities (#5118)

Signed-off-by: Stefan Weil <sw@weilnetz.de>

ThoWagen committed:
Add conditions to notices (#5087)

meganschanz committed:
Update Syndetics URL to use CDN URL instead (#5112)

demiankatz committed:
Trivial: fix typo.

EreMaijala committed:
Add configuration extraction support to the autowiring factory. (#5119)

meganschanz committed:
Update sort settings on Folio.ini for checkout items page (#5121)

ckaz committed:
Update themes/bootstrap5/templates/ContentBlock/FacetList.phtml

uncomments TwoColums version

Co-authored-by: Demian Katz <demian.katz@villanova.edu>

ckaz committed:
Update config/vufind/facets.ini

add hierarchy top title to ini

Co-authored-by: Demian Katz <demian.katz@villanova.edu>

ckaz committed:
Update module/VuFind/src/VuFind/ContentBlock/FacetList.php

uses ExplodeSettingTrait as suggested

Co-authored-by: Demian Katz <demian.katz@villanova.edu>

ckaz committed:
Adds rebuilt CSS files

ckaz commented:
> Also, I notice there is a merge conflict here that can likely be resolved by merging the dev branch, rebuilding the CSS, and committing the resulting files to clean up the conflict. If you need any help sorting that out, let me know and I'd be happy to take care of it for you.

Apologies for the delay again: 

I've just pulled the dev branch resulting in hundreds of commits in this PR. Hope I didn't mess this up. If I did, let me know and I'll build a clean PR from scratch.

demiankatz reviewed (changes requested):
@ckaz, unfortunately, something strange seems to have happened here. It seems that recent commits from dev have been added to your branch, but Git doesn't recognize it as having been merged with the dev branch. Is it possible that you did a rebase instead of a merge?

In any case, as you suggest, it may be easier to close this and start a fresh PR. Sorry for the inconvenience!

ckaz commented:
You're right. I close. Will be superseded by #5130
